src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/TaskCompletionSourceWithCancellation.cs

```diff
@@ -0,0 +1,32 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http
+{
+    internal sealed class TaskCompletionSourceWithCancellation<T> : TaskCompletionSource<T>
+    {
+        private CancellationToken _cancellationToken;
+
+        public TaskCompletionSourceWithCancellation() : base(TaskCreationOptions.RunContinuationsAsynchronously)
+        {
+        }
+
+        private void OnCancellation()
+        {
+            TrySetCanceled(_cancellationToken);
+        }
+
+        public async Task<T> WaitWithCancellationAsync(CancellationToken cancellationToken)
+        {
+            _cancellationToken = cancellationToken;
+            using (cancellationToken.Register(s => ((TaskCompletionSourceWithCancellation<T>)s).OnCancellation(), this))
```

Comment:
Nit: Register => UnsafeRegister.  The latter doesn't flow ExecutionContext into the callback; we don't need the ExecutionContext captured/flowed here (we control all of the code in the callback and know exactly what it does), and avoiding it is a small optimization.

Reply:
Thanks. This should be fixed in the latest commit.